Flaky-test: ShadowManagedLedgerImplTest.testShadowWrites

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Example failure

https://github.com/apache/pulsar/actions/runs/8415552736/job/23056991961?pr=22335

### Exception stacktrace

<!-- copy-paste the stack trace in the code block below -->
```
Error:  org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.testShadowWrites  Time elapsed: 10.183 s  <<< FAILURE!
  org.awaitility.core.ConditionTimeoutException: Assertion condition defined as a org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest expected [5:1] but found [6:0] within 10 seconds.
  	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
  	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
  	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
  	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
  	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
  	at org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.testShadowWrites(ShadowManagedLedgerImplTest.java:88)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:840)
  Caused by: java.lang.AssertionError: expected [5:1] but found [6:0]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1577)
  	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
  	at org.testng.Assert.assertEquals(Assert.java:131)
  	at org.testng.Assert.assertEquals(Assert.java:643)
  	at org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.lambda$testShadowWrites$1(ShadowManagedLedgerImplTest.java:88)
  	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
  	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:248)
  	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:235)
  	... 4 more
```


### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

another one: https://github.com/apache/pulsar/actions/runs/8532427225/job/23373708401?pr=22393#step:11:2351
something doesn't make sense "did not expect [5:1] but found [5:1]"

```
  Error:  org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.testShadowWrites  Time elapsed: 0.022 s  <<< FAILURE!
  java.lang.AssertionError: did not expect [5:1] but found [5:1]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failEquals(Assert.java:1581)
  	at org.testng.Assert.assertNotEqualsImpl(Assert.java:156)
  	at org.testng.Assert.assertNotEquals(Assert.java:2197)
  	at org.testng.Assert.assertNotEquals(Assert.java:2219)
  	at org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.testShadowWrites(ShadowManagedLedgerImplTest.java:81)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:840)
```

most recent error at https://github.com/apache/pulsar/actions/runs/8723104575/job/23931059422?pr=22524#step:11:2351
```
  Error:  org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.testShadowWrites  Time elapsed: 0.246 s  <<< FAILURE!
  java.lang.AssertionError: expected [4] but found [5]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1577)
  	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
  	at org.testng.Assert.assertEquals(Assert.java:131)
  	at org.testng.Assert.assertEquals(Assert.java:1418)
  	at org.testng.Assert.assertEquals(Assert.java:1382)
  	at org.testng.Assert.assertEquals(Assert.java:1428)
  	at org.apache.bookkeeper.mledger.impl.ShadowManagedLedgerImplTest.testShadowWrites(ShadowManagedLedgerImplTest.java:111)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:840)
```

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ShadowManagedLedgerImplTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ShadowManagedLedgerImplTest.java
@@ -51,7 +51,7 @@ public class ShadowManagedLedgerImplTest extends MockedBookKeeperTestCase {
         return (ShadowManagedLedgerImpl) shadowML;
     }
 
-    @Test(groups = "flaky")
+    @Test
     public void testShadowWrites() throws Exception {
         ManagedLedgerImpl sourceML = (ManagedLedgerImpl) factory.open("source_ML", new ManagedLedgerConfig()
                 .setMaxEntriesPerLedger(2)
@@ -76,16 +76,13 @@ public class ShadowManagedLedgerImplTest extends MockedBookKeeperTestCase {
         //Add new data to source ML
         Position newPos = sourceML.addEntry(data);
 
-        // The state should not be the same.
-        log.info("Source.LCE={},Shadow.LCE={}", sourceML.lastConfirmedEntry, shadowML.lastConfirmedEntry);
-        assertNotEquals(sourceML.lastConfirmedEntry, shadowML.lastConfirmedEntry);
-
         //Add new data to source ML, and a new ledger rolled
-        newPos = sourceML.addEntry(data);
-        assertEquals(sourceML.ledgers.size(), 4);
-        Awaitility.await().untilAsserted(()->assertEquals(shadowML.ledgers.size(), 4));
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(sourceML.ledgers.size(), 4);
+            assertEquals(shadowML.ledgers.size(), 4);
+            assertEquals(sourceML.lastConfirmedEntry, shadowML.lastConfirmedEntry);
+        });
         log.info("Source.LCE={},Shadow.LCE={}", sourceML.lastConfirmedEntry, shadowML.lastConfirmedEntry);
-        Awaitility.await().untilAsserted(()->assertEquals(sourceML.lastConfirmedEntry, shadowML.lastConfirmedEntry));
 
         {// test write entry with ledgerId < currentLedger
             CompletableFuture<Position> future = new CompletableFuture<>();
@@ -146,10 +143,10 @@ public class ShadowManagedLedgerImplTest extends MockedBookKeeperTestCase {
             }, fakePos);
             //This write will be queued unit new ledger is rolled in source.
 
-            newPos = sourceML.addEntry(data); // new ledger rolled.
-            newPos = sourceML.addEntry(data);
+            sourceML.addEntry(data); // new ledger rolled.
+            sourceML.addEntry(data);
             Awaitility.await().untilAsserted(() -> {
-                assertEquals(shadowML.ledgers.size(), 6);
+                assertEquals(shadowML.ledgers.size(), 5);
                 assertEquals(shadowML.currentLedgerEntries, 0);
             });
             assertEquals(future.get(), fakePos);

